### ISSUE TITLE
forward_kld with weighted training samples

### ISSUE THREAD

**Issue by alexmiltenberger** [VincentStimper/normalizing-flows#22]
Love this library, thanks so much @VincentStimper for your efforts in putting this together. 

I have been working lately with weighted samples and I figured out a simple change to the loss to allow for this. There are a couple ways it could be implemented, but one suggestion is to change [this method](https://github.com/VincentStimper/normalizing-flows/blob/52f2dbfd8a68f2f00f227298c3f58a1914ff618a/normflows/core.py#L87) to something like this

```
def forward_kld(self, x, weights=None):  # adds optional weight argument
    """Estimates forward KL divergence, see [arXiv 1912.02762](https://arxiv.org/abs/1912.02762)

    Args:
      x: Batch sampled from target distribution

    Returns:
      Estimate of forward KL divergence averaged over batch
    """
    log_q = torch.zeros(len(x), device=x.device)
    z = x
    for i in range(len(self.flows) - 1, -1, -1):
        z, log_det = self.flows[i].inverse(z)
        log_q += log_det
    log_q += self.q0.log_prob(z)
    
    if weights is not None:       # Apply weights if passed as an arg 
        log_q = log_q * weights

    return -torch.mean(log_q)
```

I've implemented this on a fork and it works well. Happy to provide the math if needed. It's a simple modification based on importance sampling. 

Any thoughts? Otherwise I may just open up a PR for these simple changes.

**Comment by VincentStimper** [comment]
Hi @alexmiltenberger,

nice to hear that you enjoy the package.
Instead of modifying the method, you could also just compute `log_q` using [this method](https://github.com/VincentStimper/normalizing-flows/blob/52f2dbfd8a68f2f00f227298c3f58a1914ff618a/normflows/core.py#L182) and then applying the weights when computing your loss in an extra step. This might be more efficient since in an importance sampling context the weights might depend on `log_q` as well, so you can directly reuse the results and do not have to iterate through the flow twice by calling the `forward_kld` and the `log_prob` method.
What do you think?

Best regards,
Vincent

**Comment by alexmiltenberger** [comment]
Ah, didn't realize the `log_prob` method was almost identical to `forward_kld`. This is a much better solution than modifying `forward_kld`. I just tried it and it works well. Thanks so much for getting back to me!
